src/nvim/event/socket.c

```diff
@@ -168,3 +169,71 @@ static void close_cb(uv_handle_t *handle)
     watcher->close_cb(watcher, watcher->data);
   }
 }
+
+static void connect_cb(uv_connect_t *req, int status)
+{
+  int *ret_status = req->data;
+  *ret_status = status;
+}
+
+bool socket_connect(Loop *loop, Stream *stream,
+                    bool tcp, const char *address,
+                    int timeout, const char **error)
+{
+  bool success = false;
+  int status;
+  uv_connect_t req;
+  req.data = &status;
+  uv_stream_t *uv_stream;
+  uv_getaddrinfo_t addr_req;
+  addr_req.addrinfo = NULL;
+  const struct addrinfo *addrinfo = NULL;
+  char *addr = NULL;
+  if (tcp) {
+    addr = xstrdup(address);
+    char *host_end = xstrchrnul(addr, ':');
+    if (*host_end != ':') {
+      *error = _("tcp address must be host:port");
+      goto cleanup;
+    }
+    *host_end = NUL;
+
+    const struct addrinfo hints = { .ai_family = AF_UNSPEC,
+                                    .ai_flags  = AI_NUMERICSERV };
```

Comment:
If we use tcp for a connection, we could also specify `.ai_socktype = SOCK_STREAM` to reduce the number of retries.